ddprof-lib/src/main/cpp/otel_process_ctx.cpp

```diff
@@ -0,0 +1,510 @@
+  atomic_thread_fence(memory_order_seq_cst);
+
+  // Step: Populate the signature into the mapping
+  // The signature must come last and not be reordered with the fields above by the compiler. After this step, external readers
+  // can read the signature and know that the payload is ready to be read.
+  memcpy(published_state.mapping->otel_process_ctx_signature, "OTEL_CTX", sizeof(published_state.mapping->otel_process_ctx_signature));
+
+  // Step: Change permissions on the mapping to only read permission
+  // We've observed the combination of anonymous mapping + a given number of pages + read-only permission is not very common,
+  // so this is left as a hint for when running on older kernels and the naming the mapping feature below isn't available.
+  // For modern kernels, doing this is harmless so we do it unconditionally.
+  if (mprotect(published_state.mapping, mapping_size, PROT_READ) == -1) {
+    if (otel_process_ctx_drop_current()) {
+      return (otel_process_ctx_result) {.success = false, .error_message = "Failed to change permissions on mapping (" __FILE__ ":" ADD_QUOTES(__LINE__) ")"};
+    } else {
+      return (otel_process_ctx_result) {.success = false, .error_message = "Failed to drop previous context (" __FILE__ ":" ADD_QUOTES(__LINE__) ")"};
+    }
+  }
+
+  // Step: Name the mapping so outside readers can:
+  // * Find it by name
+  // * Hook on prctl to detect when new mappings are published
+  if (prctl(PR_SET_VMA, PR_SET_VMA_ANON_NAME, published_state.mapping, mapping_size, "OTEL_CTX") == -1) {
+    // Naming an anonymous mapping is a Linux 5.17+ feature. On earlier versions, this method call can fail. Thus it's OK
+    // for this to fail because:
+    // 1. Things that hook on prctl are still able to see this call, even though it's not supported (TODO: Confirm this is actually the case)
+    // 2. As a fallback, on older kernels, it's possible to scan the mappings and look for the "OTEL_CTX" signature in the memory itself,
+    //    after observing the mapping has the expected number of pages and permissions.
+  }
+
+  // All done!
+
+  return (otel_process_ctx_result) {.success = true, .error_message = NULL};
+}
+
+bool otel_process_ctx_drop_current(void) {
+  otel_process_ctx_state state = published_state;
+
+  // Zero out the state and make sure no operations below are reordered with zeroing
+  published_state = (otel_process_ctx_state) {.publisher_pid = 0, .mapping = NULL, .payload = NULL};
+  atomic_thread_fence(memory_order_seq_cst);
+
+  // The mapping only exists if it was created by the current process; if it was inherited by a fork it doesn't exist anymore
+  // (due to the MADV_DONTFORK) and we don't need to do anything to it.
+  if (state.mapping != NULL && state.mapping != MAP_FAILED && getpid() == state.publisher_pid) {
+    long mapping_size = size_for_mapping();
+    if (mapping_size == -1 || munmap(published_state.mapping, mapping_size) == -1) return false;
```

Comment:
confusion between `state.mapping` and `published_state.mapping`

Reply:
Really nice spot! I'll open a PR to fix this.